kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+    unsigned int rotate_now:1;
+    unsigned int rotate:1;
+    unsigned int revoke:1;
+    unsigned int create:1;
+    unsigned int ro:1;
+    char frombuf[128];
+} *kadmin_request_desc;
+
+static void
+audit_trail(kadmin_request_desc r, krb5_error_code ret)
+{
+    const char *retname = NULL;
+
+    /* Get a symbolic name for some error codes */
+#define CASE(x) case x : retname = #x; break
+    switch (ret) {
+    CASE(ENOMEM);
+    CASE(EACCES);
+    CASE(HDB_ERR_NOT_FOUND_HERE);
+    CASE(HDB_ERR_WRONG_REALM);
+    CASE(HDB_ERR_EXISTS);
+    CASE(HDB_ERR_KVNO_NOT_FOUND);
+    CASE(HDB_ERR_NOENTRY);
+    CASE(HDB_ERR_NO_MKEY);
+    CASE(KRB5KDC_ERR_BADOPTION);
+    CASE(KRB5KDC_ERR_CANNOT_POSTDATE);
+    CASE(KRB5KDC_ERR_CLIENT_NOTYET);
+    CASE(KRB5KDC_ERR_C_PRINCIPAL_UNKNOWN);
+    CASE(KRB5KDC_ERR_ETYPE_NOSUPP);
+    CASE(KRB5KDC_ERR_KEY_EXPIRED);
+    CASE(KRB5KDC_ERR_NAME_EXP);
+    CASE(KRB5KDC_ERR_NEVER_VALID);
+    CASE(KRB5KDC_ERR_NONE);
+    CASE(KRB5KDC_ERR_NULL_KEY);
+    CASE(KRB5KDC_ERR_PADATA_TYPE_NOSUPP);
+    CASE(KRB5KDC_ERR_POLICY);
+    CASE(KRB5KDC_ERR_PREAUTH_FAILED);
+    CASE(KRB5KDC_ERR_PREAUTH_REQUIRED);
+    CASE(KRB5KDC_ERR_SERVER_NOMATCH);
+    CASE(KRB5KDC_ERR_SERVICE_EXP);
+    CASE(KRB5KDC_ERR_SERVICE_NOTYET);
+    CASE(KRB5KDC_ERR_S_PRINCIPAL_UNKNOWN);
+    CASE(KRB5KDC_ERR_TRTYPE_NOSUPP);
+    CASE(KRB5KRB_ERR_RESPONSE_TOO_BIG);
+    /* XXX Add relevant error codes */
+    case 0:
+        retname = "SUCCESS";
+        break;
+    default:
+        retname = NULL;
+        break;
+    }
+
+    /* Let's save a few bytes */
+    if (retname && !strncmp("KRB5KDC_", retname, sizeof("KRB5KDC_") - 1))
+        retname += sizeof("KRB5KDC_") - 1;
+#undef PREFIX
+    heim_audit_trail((heim_svc_req_desc)r, ret, retname);
+}
+
+static krb5_log_facility *logfac;
+static pthread_key_t k5ctx;
+
+static krb5_error_code
+get_krb5_context(krb5_context *contextp)
+{
+    krb5_error_code ret;
+
+    if ((*contextp = pthread_getspecific(k5ctx)))
+        return 0;
+
+    ret = krb5_init_context(contextp);
+    /* XXX krb5_set_log_dest(), warn_dest, debug_dest */
+    if (ret == 0)
+        (void) pthread_setspecific(k5ctx, *contextp);
+    return ret;
+}
+
+static int port = -1;
+static int help_flag;
+static int daemonize;
+static int daemon_child_fd = -1;
+static int local_read_only;
+static int verbose_counter;
+static int version_flag;
+static int reverse_proxied_flag;
+static int thread_per_client_flag;
+struct getarg_strings audiences;
+static const char *cert_file;
+static const char *priv_key_file;
+static const char *cache_dir;
+static const char *realm;
+static const char *hdb;
+static const char *primary_server;
+static const char *admin_server;
+static const char *primary_admin_server;
+static const char *stash_file;
+static const char *kadmin_client_name = "keytab_service";
+static const char *kadmin_client_keytab;
+
+static krb5_error_code
+get_kadm_handle(krb5_context context, void **kadm_handle, int primary)
```

Comment:
Also write a test of local read-only HDB and neither a primary kadmin server nor a redirect URI for the primary keytab service.

Reply:
The realm needs to be an argument as the HTTP request can have a `realm=REALM` q-param.